libtransact/src/protocol/receipt.rs

```diff
@@ -61,6 +62,61 @@ impl From<StateChange> for state::StateChange {
     }
 }
 
+impl TryFrom<TransactionResult> for Vec<StateChange> {
+    type Error = String;
+
+    fn try_from(result: TransactionResult) -> Result<Self, Self::Error> {
+        match result {
+            TransactionResult::Valid { state_changes, .. } => Ok(state_changes),
+            TransactionResult::Invalid { .. } => {
+                Err("cannot convert `Invalid` transaction result to state change".into())
+            }
+        }
+    }
+}
+
+impl TryFrom<TransactionResult> for Vec<state::StateChange> {
+    type Error = String;
+
+    fn try_from(result: TransactionResult) -> Result<Self, Self::Error> {
+        let state_changes: Vec<StateChange> = result.try_into()?;
+        Ok(state_changes
+            .into_iter()
+            .map(state::StateChange::from)
+            .collect())
+    }
+}
+
+impl TryFrom<TransactionReceipt> for Vec<StateChange> {
+    type Error = String;
+
+    fn try_from(receipt: TransactionReceipt) -> Result<Self, Self::Error> {
+        let TransactionReceipt {
+            transaction_result,
+            transaction_id,
+        } = receipt;
+
+        transaction_result.try_into().map_err(|err| {
+            format!(
+                "failed to convert transaction receipt {}: {}",
```

Comment:
If we introduce an explicit error as suggested above, this code should not use the string conversion on the causing error and instead provide the full error string here. This will cut down one level of error messaging (right now we are getting fairly deep on our error strings for this chaining, so an opportunity like this to cut out a level seems good). (Really, this string construction would be done via the error struct, so this probably amounts to not chaining the errors together at all (source is irrelevant).

Reply:
The new error types do not chain